control/passivity.py

```diff
@@ -45,43 +69,180 @@ def ispassive(sys):
 
+    #we're maximizing a passivity index, include it in the cost function
+    if rho is None or nu is None:
+        c = cvx.matrix(np.append(np.array(c), -1.0))
+
+    Gs = [cvx.matrix(sys_coefficents)] + [cvx.matrix(P_coefficents)]
+    hs = [cvx.matrix(sys_constants)] + [cvx.matrix(P_constants)]
+
     # crunch feasibility solution
     cvx.solvers.options['show_progress'] = False
-    sol = cvx.solvers.sdp(c,
-                          Gs=[cvx.matrix(coefficents)],
-                          hs=[cvx.matrix(constants)])
+    sol = cvx.solvers.sdp(c, Gs=Gs, hs=hs)
+    return sol["x"]
+
+
+def get_output_fb_index(sys):
+    '''Returns the output feedback passivity (OFP) index for the input system. 
+    
+    The OFP is the largest gain that can be placed in positive feedback 
+    with a system such that the new interconnected system is passive.
+
+    Parameters
+    ----------
+    sys: An LTI system
+        System to be checked.
+
+    Returns
+    -------
+    float: 
+        The OFP index 
+    '''
+    sol = solve_passivity_LMI(sys, nu=eps)
+    if sol is None:
+        raise RuntimeError("LMI passivity problem is infeasible")
+    else:
+        return sol[-1]
+
 
-    return (sol["x"] is not None)
+def get_input_ff_index(sys):
+    '''Returns the input feedforward passivity (IFP) index for the input system. 
+    
+    The IFP is the largest gain that can be placed in negative parallel interconnection 
+    with a system such that the new interconnected system is passive.
+
+    Parameters
+    ----------
+    sys: An LTI system
+        System to be checked.
+
+    Returns
+    -------
+    float: 
+        The IFP index 
+    '''
+
+    sol = solve_passivity_LMI(sys, rho=eps)
+    if sol is None:
+        raise RuntimeError("LMI passivity problem is infeasible")
+    else:
+        return sol[-1]
+
+
+def ispassive(sys, ofp_index = 0, ifp_index = 0):
+    '''Indicates if a linear time invariant (LTI) system is passive.
+
+    Checks if system is passive with the given output feedback (OFP) and input feedforward (IFP)
+    passivity indices. Querying if the system is passive in the sense of V(x)>=0 and \\dot{V}(x) <= y.T*u, 
+    is equiavlent to the default case of ofp_index = 0, ifp_index = 0.
+
+    Note that computing the ofp_index and ifp_index for a system, then using both values simultaneously 
+    to as inputs to this function is not guaranteed to have an output of 'True' 
+    (the system might not be passive with both indices at the same time). For more details, see:
+        McCourt, Michael J., and Panos J. Antsaklis
+            "Demonstrating passivity and dissipativity using computational methods." 
+
+    Parameters
+    ----------
+    sys: An LTI system
+        System to be checked.
+    ofp_index: float
+        Output feedback passivity index.
+    ifp_index: float
+        Input feedforward passivity index.
+
+    Returns
+    -------
+    bool: 
+        The system is passive.
+    '''
```

Comment:
"equiavlent"

The math doesn't render well:

![ispassive_spyder](https://user-images.githubusercontent.com/4623504/181465137-9c293e7d-c13a-4963-a166-5ce9c28df4e8.png)

This one does not throw any linter errors at me:

```suggestion
    r"""Indicate if a linear time invariant (LTI) system is passive.

    Checks if system is passive with the given output feedback (OFP) and input
    feedforward (IFP) passivity indices.

    Parameters
    ----------
    sys: An LTI system
        System to be checked.
    ofp_index: float
        Output feedback passivity index.
    ifp_index: float
        Input feedforward passivity index.

    Returns
    -------
    bool: 
        The system is passive.

    Notes
    -----
    Querying if the system is passive in the sense of
    
    .. math:: V(x) >= 0 \land \dot{V}(x) <= y^T u

    is equivalent to the default case of `ofp_index = 0`, `ifp_index = 0`.

    Note that computing the ofp_index and ifp_index for a system, then using
    both values simultaneously to as inputs to this function is not guaranteed
    to have an output of `True` (the system might not be passive with both
    indices at the same time).
    
    For more details, see [1].

    References
    ----------
    .. [1] McCourt, Michael J., and Panos J. Antsaklis
          "Demonstrating passivity and dissipativity using computational methods." 
    """   
```


![ispassive_spyder_new](https://user-images.githubusercontent.com/4623504/181469028-f64d21c5-52ed-45a4-9161-daca14a7ece8.png)


(Yours shows me D300 D301 D401 in spyder)
https://numpydoc.readthedocs.io/en/latest/format.html
http://www.pydocstyle.org/en/6.1.1/error_codes.html

Reply:
That is much better, thanks for doing that!